File: MeepMeepTesting/src/main/java/com/example/meepmeeptesting/EditablePose.java
package com.example.meepmeeptesting;

import static com.example.meepmeeptesting.AutonVars.X_START_LEFT;
import static com.example.meepmeeptesting.AutonVars.X_START_RIGHT;
import static com.example.meepmeeptesting.AutonVars.isRed;

import static java.lang.Math.toDegrees;

import com.acmerobotics.roadrunner.geometry.Pose2d;

public class EditablePose {

    public double x, y, heading;

    public EditablePose(double x, double y, double heading) {
        this.x = x;
        this.y = y;
        this.heading = heading;
    }

    public String toString() {
        return x + ", " + y + ", " + toDegrees(heading);
    }

    public EditablePose(Pose2d pose) {
        this(pose.getX(), pose.getY(), pose.getHeading());
    }

    public EditablePose clone() {
        return new EditablePose(x, y, heading);
    }

    public EditablePose byAlliance() {
        double alliance = isRed ? 1 : -1;
        return new EditablePose(
                x,
                y * alliance,
                heading * alliance
        );
    }

    public EditablePose bySide() {
        return new EditablePose(
                x + (AutonVars.backdropSide ? 0 : X_START_LEFT - X_START_RIGHT),
                y,
                heading
        );
    }

    public EditablePose flipBySide() {
        return new EditablePose(
                AutonVars.backdropSide ? x : X_START_RIGHT + X_START_LEFT - x,
                y,
                AutonVars.backdropSide ? heading : Math.PI - heading
        );
    }

    public EditablePose byBoth() {
        return byAlliance().bySide();
    }

    public Pose2d toPose2d() {
        return new Pose2d(x, y, heading);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/control/motion/EditablePose.kt
package org.firstinspires.ftc.teamcode.control.motion

import com.acmerobotics.roadrunner.geometry.Pose2d
import org.firstinspires.ftc.teamcode.opmodes.MainAuton.X_START_LEFT
import org.firstinspires.ftc.teamcode.opmodes.MainAuton.X_START_RIGHT
import org.firstinspires.ftc.teamcode.subsystems.centerstage.Robot.backdropSide
import org.firstinspires.ftc.teamcode.subsystems.centerstage.Robot.isRed

data class EditablePose

@JvmOverloads
constructor(
    @JvmField var x: Double = 0.0,
    @JvmField var y: Double = 0.0,
    @JvmField var heading: Double = 0.0,
) {
    fun byAlliance(): EditablePose {
        val alliance: Double = if (isRed) 1.0 else -1.0
        y *= alliance
        heading *= alliance
        return this
    }

    fun bySide(): EditablePose {
        if (!backdropSide) x += X_START_LEFT - X_START_RIGHT
        return this
    }

    fun byBoth(): EditablePose {
        return byAlliance().bySide()
    }

    fun toPose2d(): Pose2d {
        return Pose2d(x, y, heading)
    }

    fun flipBySide(): EditablePose {
        if (!backdropSide) heading = Math.PI - heading
        if (!backdropSide) x = (X_START_LEFT + X_START_RIGHT) - x
        return this
    }
}